### PR TITLE
feat: Finalize enrollment WPB-6039

### PR DESCRIPTION
**Pull request by KaterinaWire**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-6039" title="WPB-6039" target="_blank">StoryWPB-6039</a>  [iOS] Finalize enrollment
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

Finish Certificate Enrollment for CC version 30.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-request-strategy/Sources/E2EIdentity/E2eIEnrollment.swift [hunk 1]
```diff
@@ -349,9 +387,18 @@ public struct ChallengeResponse: Codable, Equatable {
 
 }
 
+public struct ChallengeResponse1: Codable, Equatable {
```
Comment (typfel) [inline wire-ios-request-strategy/Sources/E2EIdentity/E2eIEnrollment.swift]: There will be a `ChallengeResponse2` later?

**Review by typfel — APPROVED** [review]
👌

File: wire-ios-request-strategy/Sources/E2EIdentity/E2eIEnrollment.swift (hunk 1 above)
Comment (KaterinaWire) [reply wire-ios-request-strategy/Sources/E2EIdentity/E2eIEnrollment.swift]: No, removed

File: wire-ios-request-strategy/Sources/E2EIdentity/E2eIEnrollment.swift [hunk 2]
```diff
@@ -318,6 +342,18 @@ public final class E2eIEnrollment: E2eIEnrollmentInterface {
         }
     }
 
+    public func getOAuthRefreshToken()  async throws -> String? {
```
Comment (David-Henner) [inline wire-ios-request-strategy/Sources/E2EIdentity/E2eIEnrollment.swift:345]: Do we expect the refresh token to ever be `nil` or empty? If it ever is, should it be considered an error? Could we return a non-optional `String` and throw an error if the token is nil or empty?

File: wire-ios-request-strategy/Sources/E2EIdentity/EnrollE2eICertificateUseCase.swift [hunk 3]
```diff
@@ -54,48 +61,66 @@ public final class EnrollE2eICertificateUseCase: EnrollE2eICertificateUseCaseInt
...
                                                                       dpopToken: dpopToken)
 
+        let refreshTokenFromCC = try? await enrollment.getOAuthRefreshToken()
+
         let dpopChallengeResponse = try await enrollment.validateDPoPChallenge(accessToken: wireAccessToken.token,
                                                                                prevNonce: authzResponse.nonce,
                                                                                acmeChallenge: wireDpopChallenge)
 
         let oidcChallengeResponse = try await enrollment.validateOIDCChallenge(idToken: idToken,
+                                                                               refreshToken: refreshTokenFromCC ?? refreshToken,
```
Comment (David-Henner) [inline wire-ios-request-strategy/Sources/E2EIdentity/EnrollE2eICertificateUseCase.swift:88]: I guess this answers my question above

File: wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift [hunk 4]
```diff
@@ -24,7 +24,10 @@ import WireRequestStrategy
 
 public protocol OAuthUseCaseInterface {
 
-    func invoke(for identityProvider: URL) async throws -> IdToken
+    func invoke(for identityProvider: URL,
+                clientID: String,
+                keyauth: String,
+                acmeAudience: String) async throws -> (idToken: String, refreshToken: String)
```
Comment (David-Henner) [inline wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift]: nitpick: formatting
```suggestion
    func invoke(
        for identityProvider: URL,
        clientID: String,
        keyauth: String,
        acmeAudience: String
    ) async throws -> (idToken: String, refreshToken: String)
```

File: wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift [hunk 5]
```diff
@@ -38,17 +41,17 @@ public class OAuthUseCase: OAuthUseCaseInterface {
         self.rootViewController = rootViewController
     }
 
-    public func invoke(for identityProvider: URL) async throws -> IdToken {
+    public func invoke(for identityProvider: URL,
+                       clientID: String,
+                       keyauth: String,
+                       acmeAudience: String) async throws -> (idToken: String, refreshToken: String) {
```
Comment (David-Henner) [inline wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift]: nitpick: formatting

```suggestion
    public func invoke(
        for identityProvider: URL,
        clientID: String,
        keyauth: String,
        acmeAudience: String
    ) async throws -> (idToken: String, refreshToken: String) {
```

File: wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift [hunk 6]
```diff
@@ -74,7 +78,32 @@ public class OAuthUseCase: OAuthUseCaseInterface {
...
+            case value = "value"
+        }
+
+        let keyauth: [String: Any] = [CodingKeys.essential.rawValue: true,
+                                      CodingKeys.value.rawValue: "\(keyauth)"]
+        let acmeAud: [String: Any] = [CodingKeys.essential.rawValue: true,
+                                       CodingKeys.value.rawValue: "\(acmeAudience)"]
+        let idToken = [CodingKeys.idToken.rawValue: [CodingKeys.keyauth.rawValue: keyauth,
+                                                       CodingKeys.acmeAud.rawValue: acmeAud]
+        ]
```
Comment (David-Henner) [inline wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/OAuthUseCase.swift]: nitpick: formatting

```suggestion
        let keyauth: [String: Any] = [
            CodingKeys.essential.rawValue: true,
            CodingKeys.value.rawValue: "\(keyauth)"
        ]
        let acmeAud: [String: Any] = [
            CodingKeys.essential.rawValue: true,
            CodingKeys.value.rawValue: "\(acmeAudience)"
        ]
        let idToken = [
            CodingKeys.idToken.rawValue: [
                CodingKeys.keyauth.rawValue: keyauth,
                CodingKeys.acmeAud.rawValue: acmeAud
            ]
        ]
```

**Review by David-Henner — APPROVED** [review]

**Pull request merged**